lib/internal/fs/promises.js

```diff
@@ -790,50 +782,56 @@ async function readdirRecursive(originalPath, options) {
         !!options.withFileTypes,
         kUsePromises,
       ),
+      0,
     ],
   ];
 
-
   if (options.withFileTypes) {
     while (queue.length > 0) {
       // If we want to implement BFS make this a `shift` call instead of `pop`
-      const { 0: path, 1: readdir } = ArrayPrototypePop(queue);
+      const { 0: path, 1: readdir, 2: depth } = ArrayPrototypePop(queue);
       for (const dirent of getDirents(path, readdir)) {
-        ArrayPrototypePush(result, dirent);
-        if (dirent.isDirectory()) {
-          const direntPath = pathModule.join(path, dirent.name);
-          ArrayPrototypePush(queue, [
-            direntPath,
-            await binding.readdir(
+        const direntPath = pathModule.join(path, dirent.name);
+        if (typeof options.filter !== 'function' || options.filter(direntPath, depth)) {
```

Comment:
```suggestion
        if (typeof options.filter !== 'function' || options.filter(dirent, depth)) {
```
IMHO if `option.withFileTypes` is truthy, it would make sense to pass `dirent` directly, wdyt?